sqlmesh/core/snapshot/definition.py

```diff
@@ -1865,6 +1919,86 @@ def snapshots_to_dag(snapshots: t.Collection[Snapshot]) -> DAG[SnapshotId]:
     return dag
 
 
+def apply_auto_restatements(
+    snapshots: t.Dict[SnapshotId, Snapshot], execution_time: TimeLike
+) -> t.List[SnapshotIntervals]:
+    """Applies auto restatements to the snapshots.
+
+    This operation results in the removal of intervals for snapshots that are ready to be restated based
+    on the provided execution time and configured auto restatement settings. For each affected snapshot,
+    it also updates the next auto restatement timestamp.
+
+    Args:
+        snapshots: A dictionary of snapshots to apply auto restatements to.
+        execution_time: The execution time.
+
+    Returns:
+        A list of SnapshotIntervals with **new** intervals that need to be restated.
+    """
+    dag = snapshots_to_dag(snapshots.values())
+    auto_restated_intervals_per_snapshot: t.Dict[SnapshotId, Intervals] = {}
+    for s_id in dag:
+        if s_id not in snapshots:
+            continue
+        snapshot = snapshots[s_id]
+
+        next_auto_restated_interval = snapshot.get_next_auto_restatement_interval(execution_time)
+        auto_restated_intervals = [
+            interval
+            for parent_s_id in snapshot.parents
+            if parent_s_id in auto_restated_intervals_per_snapshot
+            for interval in auto_restated_intervals_per_snapshot[parent_s_id]
+        ]
+        if next_auto_restated_interval:
+            logger.info(
+                "Calculated the next auto restated interval (%s, %s) for snapshot %s",
+                time_like_to_str(next_auto_restated_interval[0]),
+                time_like_to_str(next_auto_restated_interval[1]),
+                snapshot.snapshot_id,
+            )
+            auto_restated_intervals.append(next_auto_restated_interval)
+
+        normalized_auto_restated_intervals = []
+        for interval in merge_intervals(auto_restated_intervals):
+            interval_start = snapshot.node.interval_unit.cron_floor(interval[0])
+            interval_end = snapshot.node.interval_unit.cron_floor(interval[1])
```

Comment:
what is this section doing?

Reply:
If the upstream model is hourly and the downstream is daily, this expands the interval to restate a complete interval for the downstream model.
If it's `01/02 10:00:00` right now, and we restate 24 intervals of the hourly model we need to make sure we restate the whole `01/01` for the daily model. This handles the partial interval